tests/canopen/test_canopen_network.py

```diff
@@ -19,3 +70,50 @@ def test_scan_slaves(read_config):
     slaves = net.scan_slaves()
     assert len(slaves) > 0
 
+
+@pytest.mark.no_connection
+def test_scan_slaves_none_nodes(virtual_network):
+    nodes = virtual_network.scan_slaves()
+    assert len(nodes) == 0
+
+
+@pytest.mark.canopen
+def test_disconnect_from_slave(read_config):
+    protocol_contents = read_config['canopen']
+    net = CanopenNetwork(device=CAN_DEVICE(protocol_contents['device']),
+                         channel=protocol_contents['channel'],
+                         baudrate=CAN_BAUDRATE(protocol_contents['baudrate']))
+
+    servo = net.connect_to_slave(
+        target=protocol_contents['node_id'],
+        dictionary=protocol_contents['dictionary'],
+        eds=protocol_contents['eds'])
+
+    assert len(net.servos) == 1
+    net.disconnect_from_slave(servo)
+    assert len(net.servos) == 0
+
+
+@pytest.mark.no_connection
+def test_setup_and_teardown_connection(virtual_network):
+    assert virtual_network._connection is None
+    virtual_network._setup_connection()
+    assert virtual_network._connection is not None
+    virtual_network._teardown_connection()
+    assert virtual_network._connection is None
+
+
+@pytest.mark.skip
+def test_load_firmware(connect_to_slave, read_config):
+    # TODO: Fix load_firmware method to work independently of status listeners
+    servo, net = connect_to_slave
+    assert servo is not None and net is not None
+    assert len(net.servos) == 1
+    fw_version = servo.read('DRV_ID_SOFTWARE_VERSION')
+    protocol_contents = read_config['canopen']
+
+    net.load_firmware(protocol_contents['node_id'], protocol_contents['test_fw_file'])
+    new_fw_version = servo.read('DRV_ID_SOFTWARE_VERSION')
+    
+    assert new_fw_version != fw_version
```

Comment:
We have to make sure that the current firmware version is different than the firmware version to be loaded.

Reply:
Furthermore, do we want to switch back to the previous firmware after this test?